File: wk01/src/zalg.py

def zalg(string: str) -> list[int]:
    n = len(string) # 18
    z = [0] * n
    z[0] = n
    print('set 0th index to n')

    left, right = -1, -1

    for k in range(1,n):
        k_mirror = k - left
        zbox_offset = right - k
        
        # case 1 (k > r)
        if k > right:
            print('case 1')
            i = 0
            while k + i < n and string[i] == string[k + i]:
                i += 1
            z[k] = i
            if z[k] > 0:
                left = k
                right = k + z[k]
        # case 2 (k <= r)
        elif k <= right:
            z[k] = min(z[k_mirror], right - k)

            if z[k_mirror] == zbox_offset:
                i = 0
                while right + i < n and string[right + i] == string[right - k + i]:
                    i += 1
                z[k] += i
                if z[k] > 0:
                    left = k
                    right = k + z[k]

            # # 2a: fully contained within zbox
            # if z[k_mirror] < zbox_offset:
            #     print('case 2a: reused previous value')
            #     z[k] = z[k_mirror]

            # # 2b
            # elif z[k_mirror] >= zbox_offset:
            #     print('case 2b + possibly 2c')
            #     z[k] = right - k
            #     # 2c: extend past right of zbox
            #     if z[k_mirror] == zbox_offset:
            #         i = 0
            #         while right + i < n and string[right + i] == string[right - k + i]:
            #             i += 1
            #         z[k] += i
            #         if z[k] > 0:
            #             left = k
            #             right = k + z[k]

    
    return z

if __name__ == '__main__':
    z = zalg("aabcaabxaabcaabcay")
    print(z)

    correct = [18,1,0,0,3,1,0,0,7,1,0,0,5,1,0,0,1,0]
    assert z[0] == correct[0]
    assert z == correct, f'\nExpected {correct},\nGot      {z}'

    del z
    del correct
    print('\n')

    z = zalg("abaaba")
    print(z)
    
    correct = [6,0,1,3,0,1]
    assert z == correct, f'\nExpected {correct},\nGot      {z}'

    del z
    del correct
    print("\n")

    z = zalg("aaaaaaaa")
    print(z)

File: wk01/src/zalg.c
#include <stdio.h>

int main() {
    const char string[] = "abaaba";
    const size_t n = sizeof(string) - 1; // 6 elements

    int z[n];
    for (size_t i = 0; i < n; i++) z[i] = 0;
    z[0] = n;

    int left = -1;
    int right = -1;

    // naive
    for (int k = 1; k < n; k++) {
        if (k > right) {
            // case 1: naive comparison
            size_t j = 0;
            while (k+j < n && string[j] == string[k+j]) {
                j++;
            }
            z[k] = j;
            if (z[k] > 0) {
                left = k;
                right = k+z[k];
            }
        } 
        else if (k <= right) {
            // case 2a
            if (z[k-left] < right-k) {
                z[k] = z[k-left];
            }
            // case 2b
            else if (z[k-left] >= right-k) {
                z[k] = right-k;

                // case 2c
                if (z[k-left] == right-k) {
                    size_t m = 0;
                    while (right+m < n && string[right+m] == string[right-k+m]) {
                        m++;
                    }
                    z[k] += m;
                    if (z[k] > 0) {
                        left = k;
                        right = k+z[k];
                    }
                }
            }
        }
    }

    // print array
    printf("[");
    for (int m = 0; m < n; m++) {
        printf("%d", z[m]);
        if (m < n - 1) { printf(", "); }
    }
    printf("]");

    return 0;
}

